### DISCUSSION TITLE
About watermarks of physical memory allocation

### DISCUSSION THREAD

**Discussion by LinPoly** [vllm-project/vllm#263] [Q&A]
In current verison of codes(0.1.1), I noticed that both the [can_allocate()](https://github.com/vllm-project/vllm/blob/main/vllm/core/block_manager.py#L80) method and [can_swap_in()](https://github.com/vllm-project/vllm/blob/main/vllm/core/block_manager.py#L158) method of the BlockSpaceManager class deal with watermarks while [can_append_slot()](https://github.com/vllm-project/vllm/blob/main/vllm/core/block_manager.py#L105) doesn't. It seems that they should have the same mechanism on GPU memory management.

**Comment by WoosukKwon** [comment]
@LinPoly Thanks for the question! This is not a bug. The watermark is to prevent frequent preemptions (i.e., swapping or recomputation) which can be caused by accepting too many **new** requests in the batch. For the existing requests in the batch, we want them to use every slot in the KV cache.
